### PR TITLE
es & cqrs

### LINKED ISSUE #16: event sourcing & CQRS

### PR DESCRIPTION
**Pull request by eyakauleva**
close #16

### CODE REVIEW

File: src/main/java/com/solvd/micro9/tickets/domain/es/Es.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import org.springframework.data.annotation.Id;
+import org.springframework.data.mongodb.core.mapping.Field;
+
+import java.time.LocalDateTime;
+
+@Data
+@AllArgsConstructor
+public class Es {
```
Comment (h1alexbel) [inline src/main/java/com/solvd/micro9/tickets/domain/es/Es.java:18]: what is `ES`?

File: src/main/java/com/solvd/micro9/tickets/persistence/snapshot/EventRepository.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package com.solvd.micro9.tickets.persistence.snapshot;
+
+import com.solvd.micro9.tickets.domain.aggregate.Event;
+import org.springframework.data.r2dbc.repository.R2dbcRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
```
Comment (h1alexbel) [inline src/main/java/com/solvd/micro9/tickets/persistence/snapshot/EventRepository.java]: redundant

File: src/main/java/com/solvd/micro9/tickets/service/impl/EsEventCommandHandlerImpl.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+    @SneakyThrows
+    @Override
+    public Mono<EsEvent> apply(CreateEventCommand command) {
+        ObjectMapper mapper = new ObjectMapper();
+        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd hh:mm");
+        mapper.registerModule(new JavaTimeModule());
+        mapper.setDateFormat(dateFormat);
+        String payload = mapper.writeValueAsString(command.getEvent());
+
```
Comment (h1alexbel) [inline src/main/java/com/solvd/micro9/tickets/service/impl/EsEventCommandHandlerImpl.java]: empty lines means code smell

File: src/main/java/com/solvd/micro9/tickets/service/impl/EsEventCommandHandlerImpl.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        EsEvent event = EsEvent.builder()
+                .type(EsEventType.EVENT_CREATED)
+                .time(LocalDateTime.now())
+                .createdBy(command.getCommandBy())
+                .entityId(UUID.randomUUID().toString())
+                .payload(payload)
+                .build();
+
```
Comment (h1alexbel) [inline src/main/java/com/solvd/micro9/tickets/service/impl/EsEventCommandHandlerImpl.java]: empty lines means code smell

File: src/main/java/com/solvd/micro9/tickets/service/impl/EsTicketCommandHandlerImpl.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+                .zipWith(esTicketRepository.save(event))
+                .map(Tuple2::getT2)
+                .doOnSuccess(esTicket -> producer.send("New event", esTicket));
+    }
+
+    @Override
+    public void apply(SetTicketsUserIdToNullByUserIdCommand command) {
+        List<EsTicket> esTicketList = new ArrayList<>();
+        final boolean[] isStreamCompleted = {false};
+
```
Comment (h1alexbel) [inline src/main/java/com/solvd/micro9/tickets/service/impl/EsTicketCommandHandlerImpl.java]: empty lines

File: src/main/java/com/solvd/micro9/tickets/service/impl/EsTicketCommandHandlerImpl.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+                            .entityId(esTicket.getEntityId())
+                            .payload(TICKET_USER_ID_TO_NULL_PAYLOAD)
+                            .build();
+                    esTicketList.add(event);
+                })
+                .doOnComplete(() -> isStreamCompleted[0] = true)
+                .subscribe();
+
+        while (!isStreamCompleted[0]) {
+        } //TODO is there a better way to do?
```
Comment (h1alexbel) [inline src/main/java/com/solvd/micro9/tickets/service/impl/EsTicketCommandHandlerImpl.java:86]: @eyakauleva I will pull it and report to you

File: src/main/java/com/solvd/micro9/tickets/persistence/snapshot/EventRepository.java (hunk 2 above)
Comment (eyakauleva) [reply src/main/java/com/solvd/micro9/tickets/persistence/snapshot/EventRepository.java]: fixed 1ff3330

File: src/main/java/com/solvd/micro9/tickets/service/impl/EsEventCommandHandlerImpl.java (hunk 3 above)
Comment (eyakauleva) [reply src/main/java/com/solvd/micro9/tickets/service/impl/EsEventCommandHandlerImpl.java]: fixed 1ff3330

File: src/main/java/com/solvd/micro9/tickets/service/impl/EsEventCommandHandlerImpl.java (hunk 4 above)
Comment (eyakauleva) [reply src/main/java/com/solvd/micro9/tickets/service/impl/EsEventCommandHandlerImpl.java]: fixed 1ff3330

File: src/main/java/com/solvd/micro9/tickets/service/impl/EsTicketCommandHandlerImpl.java (hunk 5 above)
Comment (eyakauleva) [reply src/main/java/com/solvd/micro9/tickets/service/impl/EsTicketCommandHandlerImpl.java]: fixed 1ff3330

File: src/main/java/com/solvd/micro9/tickets/domain/es/Es.java (hunk 1 above)
Comment (eyakauleva) [reply src/main/java/com/solvd/micro9/tickets/domain/es/Es.java:18]: java doc added 1ff3330

File: src/main/java/com/solvd/micro9/tickets/service/impl/EsTicketCommandHandlerImpl.java (hunk 6 above)
Comment (h1alexbel) [reply src/main/java/com/solvd/micro9/tickets/service/impl/EsTicketCommandHandlerImpl.java:86]: @eyakauleva try to use `flatMap` or `handle`?
https://projectreactor.io/docs/core/release/reference/#_handle

```java
esTicketRepository.findAll()
.handle() // filtering and mapping
.subscribeOn(Schedulers.boundElastic())
                                 ...

```

Comment (eyakauleva) [reply src/main/java/com/solvd/micro9/tickets/service/impl/EsTicketCommandHandlerImpl.java:86]: @h1alexbel flatMap not working

just tried with handle and that's not working neither

option 1:
```java
public void apply(SetTicketsUserIdToNullByUserIdCommand command) {
        Flux<EsTicket> esTicketFlux = esTicketRepository.findAll()
                .handle((esTicket, sink) -> {
                    Ticket ticket = new Gson().fromJson(esTicket.getPayload(), Ticket.class);
                    EsTicket event = EsTicket.builder()
                            .type(EsType.TICKET_USER_DELETED)
                            .time(LocalDateTime.now())
                            .createdBy(command.getCommandBy())
                            .entityId(esTicket.getEntityId())
                            .build();
                    if(esTicket.getType().equals(EsType.TICKET_CREATED)
                            && command.getUserId().equals(ticket.getUserId())) {
                        sink.next(event);
                    }
                });
        esTicketFlux.subscribe(esTicket -> {
            esTicketRepository.save(esTicket).subscribe();
        });
    }
```
option 2:
```java
public void apply(SetTicketsUserIdToNullByUserIdCommand command) {
        esTicketRepository.findAll()
                .handle((esTicket, sink) -> {
                    Ticket ticket = new Gson().fromJson(esTicket.getPayload(), Ticket.class);
                    EsTicket event = EsTicket.builder()
                            .type(EsType.TICKET_USER_DELETED)
                            .time(LocalDateTime.now())
                            .createdBy(command.getCommandBy())
                            .entityId(esTicket.getEntityId())
                            .build();
                    if(esTicket.getType().equals(EsType.TICKET_CREATED)
                            && command.getUserId().equals(ticket.getUserId())) {
                       esTicketRepository.save(event).subscribe();
                    }
                })
                .subscribe();
    }
```

Comment (h1alexbel) [reply src/main/java/com/solvd/micro9/tickets/service/impl/EsTicketCommandHandlerImpl.java:86]: @eyakauleva forked, I'll create an issue

**Pull request merged**